pkg/controllers/jobset_controller.go

```diff
@@ -496,6 +568,15 @@ func isJobFinished(job *batchv1.Job) (bool, batchv1.JobConditionType) {
 	return false, ""
 }
 
+func jobSetSuspended(js *jobset.JobSet) bool {
```

Comment:
If we can remove it above then we can remove but currently we use it to detect difference between a normal job and suspended job being resumed.

Reply:
Given the current logic I think it is necessary. Checking `jobset.spec.suspend` and the jobset status condition of `JobSetSuspended` are used to determine which point in the suspend/resume logic we are currently in, so I don't see how we could remove it. What were you thinking

Either way I think this is fine for now and we can look into this further in a follow up PR if necessary.

Reply:
I don't get why it is necessary to check? `spec.suspend` is the source of truth that dictates how to reconcile.

Reply:
Oh I thought we were using both. In that case, I think maybe it is not necessary. To confirm, the logic should be:

Suspend:
1. user suspends jobset (`.spec.suspended=true`), reconciliation triggered
2. controller sees `.spec.suspended=true`, suspends all jobs which are not already suspended, then sets jobset status to suspended

Resume:
1. user resumes jobset (`.spec.suspended=false`), reconciliation triggered
2. controller sees `.spec.suspended=false`, then resumes all jobs which are currently suspended, then updates the jobset's `JobSetSuspended` condition to `ConditionStatus=false`.

If the above is correct, then we don't need conditional logic based on `jobSetSuspended()` check, so we can remove it.

Reply:
Third case:

JobSet starts without suspend ever set (so not resume).

The third case is why I added the check for condition.  

I thought it would be weird for a JobSet to send a condition saying the job is resumed even if the suspend flag was never set to true.  I was thinking a jobset that starts with no suspend field should not emit a condition saying it was resumed.

Reply:
ok, fair point, then I think ensureCondition should simply be changed to not set the condition if it doesn't exist and the ConditionStatus=false.

Reply:
Not sure what you are asking.  

If the condition doesn't exist, we use ensureCondition to set a condition and then update the status field.  

If a ConditionStatus=True -> ConditionStatus=False I think we would want to also capture the change.  

Or are you saying that if we suspend a condition we emit a suspended condition and then if we resume, we append a new condition that says it was resumed and keep the old suspend condition?

Reply:
> If the condition doesn't exist, we use ensureCondition to set a condition and then update the status field.

I am saying if the condition does not exist and the condition we are trying to set is associated with ConditionStatus=False, we don't set the condition at all. The condition not existing is equal to the condition exist with ConditionStatus=False

basically updateCondition will become:

```
func updateCondition(js *jobset.JobSet, condition metav1.Condition) bool {
	condition.LastTransitionTime = metav1.Now()
	for i, val := range js.Status.Conditions {
		if condition.Type == val.Type && condition.Status != val.Status {
			js.Status.Conditions[i] = condition
			// Condition found but different status so we should update
			return true
		} else if condition.Type == val.Type && condition.Status == val.Status {
			// Duplicate condition so no update
			return false
		}
	}
        // condition doesn't exist, update only if the status is false
        if condition.Status == corev1.ConditionTrue {
	  js.Status.Conditions = append(js.Status.Conditions, condition)
	  return true
        }
       return false
}```

Reply:
Ah thank you for clearing that up.  I updated it and added a few more test cases.